### PR TITLE
Modify generateExit to return full set of signing keys

### PR DESCRIPTION
**Pull request by rnicoll**
Modify generateBuild to return full set of signing keys in preparation for anonymity work meaning that owner and issuer keys are typically not the same.

### CODE REVIEW

**Comment by rnicoll** [comment]
@mikehearn can this be reviewed please? It underpins #714

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 1]
```diff
@@ -179,6 +180,23 @@ interface ServiceHub : ServicesForResolution {
     fun createSignature(signedTransaction: SignedTransaction): DigitalSignature.WithKey = createSignature(signedTransaction, legalIdentityKey)
 
     /**
+     * Helper method to create an additional signature for a [WireTransaction].
+     * @param wireTransaction The [WireTransaction] to which the signature will apply.
+     * @param publicKey The [PublicKey] matching to a signing [PrivateKey] hosted in the node.
+     * If the [PublicKey] is actually a [CompositeKey] the first leaf key found locally will be used for signing.
+     * @return The [DigitalSignature.WithKey] generated by signing with the internally held [PrivateKey].
+     */
+    fun createSignature(wireTransaction: WireTransaction, publicKey: PublicKey): DigitalSignature.WithKey = keyManagementService.sign(wireTransaction.id.bytes, publicKey)
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: @roger3cev @mnesbit  Is this compatible with where you want to take the transaction signing APIs? I guessed that ServiceHub lacks createSignature(WireTranscation) currently for a reason?

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 2]
```diff
@@ -179,6 +180,23 @@ interface ServiceHub : ServicesForResolution {
...
+
+    /**
+     * Helper method to create an additional signature for an existing (partially) SignedTransaction
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: This comment doesn't seem to match what it does.

File: finance/src/main/kotlin/net/corda/contracts/asset/Obligation.kt [hunk 3]
```diff
@@ -455,11 +455,11 @@ class Obligation<P : Any> : Contract {
      * @param amountIssued the amount to be exited, represented as a quantity of issued currency.
      * @param assetStates the asset states to take funds from. No checks are done about ownership of these states, it is
      * the responsibility of the caller to check that they do not exit funds held by others.
-     * @return the public key of the assets issuer, who must sign the transaction for it to be valid.
+     * @return the public keys who must sign the transaction for it to be valid.
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/asset/Obligation.kt]: who -> which

File: finance/src/main/kotlin/net/corda/contracts/asset/OnLedgerAsset.kt [hunk 4]
```diff
@@ -226,11 +228,11 @@ abstract class OnLedgerAsset<T : Any, C : CommandData, S : FungibleAsset<T>> : C
      * necessarily owned by us.
      * @param assetStates the asset states to take funds from. No checks are done about ownership of these states, it is
      * the responsibility of the caller to check that they do not exit funds held by others.
-     * @return the public key of the assets issuer, who must sign the transaction for it to be valid.
+     * @return the public keys who must sign the transaction for it to be valid.
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/asset/OnLedgerAsset.kt]: who -> which

File: finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt [hunk 5]
```diff
@@ -26,15 +26,15 @@ abstract class AbstractConserveAmount<S : FungibleAsset<T>, C : CommandData, T :
      * @param amountIssued the amount to be exited, represented as a quantity of issued currency.
      * @param assetStates the asset states to take funds from. No checks are done about ownership of these states, it is
      * the responsibility of the caller to check that they do not attempt to exit funds held by others.
-     * @return the public key of the assets issuer, who must sign the transaction for it to be valid.
+     * @return the public keys who must sign the transaction for it to be valid.
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: who -> which

File: finance/src/main/kotlin/net/corda/flows/CashExitFlow.kt [hunk 6]
```diff
@@ -62,7 +62,9 @@ class CashExitFlow(val amount: Amount<Currency>, val issueRef: OpaqueBytes, prog
                 .toSet()
         // Sign transaction
         progressTracker.currentStep = SIGNING_TX
-        val tx = serviceHub.signInitialTransaction(builder)
+        val wtx = builder.toWireTransaction()
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/flows/CashExitFlow.kt]: I'm not sure this is the right API. Again, @mnesbit would like your views here?

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt (hunk 1 above)
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: I agree signing WireTransactions sounds wrong. There is already a method that takes multiple keys to seal a TransactionBuilder.

File: finance/src/main/kotlin/net/corda/flows/CashExitFlow.kt (hunk 6 above)
Comment (mnesbit) [inline finance/src/main/kotlin/net/corda/flows/CashExitFlow.kt]: Yes I would like to seal the builder in one step and I think there is already an API that can do this. We may need to filter out keys that are ours, versus those that are other peoples. However, I would have felt we can determine this from the keys set on the KeyManagementService, or better yet by context/deanonymised Party data. My ideal is that we never use anonymised Party entries in flows and there is a helper for TransactionBuilders to subtitute anonymous keys via Kryo at the point we freeze and sign. Then on message receipt we should be able to use a different Kryo to de-anonymise the Parties. We did discuss this with @mikehearn originaly, and I still think it is a valid way of making this stuff invisible to the flow user.

**Comment by rnicoll** [comment]
Removed the functions added to `ServiceHub`, having had it pointed out that there's a `signInitialTransaction` that takes in a list of keys. Modified`signInitialTransaction` to take in any iterable rather than only lists, too.

**Pull request merged**